scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
+    # sort on operation
+    possible_permissions.sort(key=lambda p: p[3])
+    # sort on topic names, because that looks nice for users, but since there is a bijection 
+    # between topic names and IDs this will place all matching topic IDs together in blocks 
+    # in some order
+    possible_permissions.sort(key=lambda p: p[2])
+
+    def equivalent(p1, p2):
+        return p1[1] == p2[1] and p1[3]==p2[3];
+    
+    # lack of an obvious analogue to std::unique makes this awkward, and not in-place
+    dedup = []
+    last = None
+    for p in possible_permissions:
+        if last is None or not equivalent(last,p):
+            dedup.append(p)
+            last=p
+    
+    return dedup
+
+
+# Find a new, non-excluded group permission which can support the given 
+# credential permission (permission), or delete it if none can be found.  
+# exclude is a callable which will be called with each relevant group permission 
+# whose return value should indicate whether it is to be ignored or not.
+# user_memberships should be the set of group memberships for the user who owns
+# permission
+# group_permissions is a cache of the permissions belonging to groups of which
+# the user is a member. It will be automatically updated. 
+def repairOrDeletePermission(permission, exclude, user_memberships, group_permissions):
+    amended = False
+    for membership in user_memberships:
+        # fetch permissions for this group if not already cached
+        if not membership.group_id in group_permissions:
+            group_permissions[membership.group_id] = GroupKafkaPermission.objects.filter(principal=membership.group_id)
+        for group_perm in group_permissions[membership.group_id]:
+            if exclude(group_perm):
+                continue # this is one of the permissions we cannot use
+            if supportingPermission(group_perm, permission):
+                # this other group permission is a valid substitute for the one which is 
+                #going away, so we can correct the credential permission instead of removing it
+                permission.parent = group_perm
+                permission.save()
+                amended = True
+                break
+        if amended:
+            break # if we fixed this permission we can stop working on it
+    if not amended:
+        # if we could not find a way to fix the permission we must delete it
+        permission.delete()
+    return amended
+
+
+# remove all permissions the given user recieved via the given group, either 
+# because the user is being removed from that group or the group is being
+# deleted entirely, retaining credential permissions if they can be equivalently 
+# constructed via another group permission
+def removeUserGroupPermissions(user_id, group_id):
+    # must check all credentials owned by this user
+    user_creds = SCRAMCredentials.objects.filter(owner=user_id)
+    print("User",user_id,"has",len(user_creds),"credentials")
```

Comment:
Use `logging` instead of printing, and this will go to the right places. See eg https://github.com/scimma/scimma-admin/blob/c8f3a1ad64cb217f43b8ba4c0d0dc0650ab62b72/scimma_admin/hopskotch_auth/auth.py#L27.

Reply:
All remaining uses of `print` were temporary debugging code which I will just delete; `logging` is already used for all actual logging.